src/lib/awesomeBarObserver.js

```diff
@@ -0,0 +1,528 @@
+/* global FrecencyOptimizer */
+/* eslint no-unused-vars: ["error", { "varsIgnorePattern": "(svmLoss|AwesomeBarObserver)" }]*/
+
+class AwesomeBarObserver {
+  constructor(synchronizer) {
+    this.frecencyOptimizer = new FrecencyOptimizer(
+      synchronizer,
+      FrecencyOptimizer.svmLoss,
+    );
+    this.eventsToObserve = [
+      "onFocus",
+      "onBlur",
+      "onKeyDown",
+      "onKeyPress",
+      "onInput",
+      "onAutocompleteSuggestionsHidden",
+      "onAutocompleteSuggestionsUpdated",
+      "onAutocompleteSuggestionSelected",
+    ];
+    this.observedEventsSinceLastFocus = [];
+  }
+  async start() {
+    this.eventsToObserve.map(eventRef => {
+      browser.experiments.awesomeBar[eventRef].addListener(
+        this[eventRef].bind(this),
+      );
+    });
+    await browser.experiments.awesomeBar.start();
+  }
+
+  async stop() {
+    await browser.experiments.awesomeBar.stop();
+    this.eventsToObserve.map(eventRef => {
+      browser.experiments.awesomeBar[eventRef].removeListener(this[eventRef]);
+    });
+  }
+
+  /**
+   * Three main interactions with the awesome bar trigger a model update via study telemetry:
+   * 1. A suggestion was selected from the autocomplete popup
+   * 2. The autocomplete popup got some suggestions displayed but none were selected
+   * 3. The autocomplete popup did not get some suggestions displayed and none was selected
+   * @returns {Promise<boolean>} Promise that resolves when the method has completed
+   */
+  async afterInteraction() {
+    await browser.study.logger.debug([
+      "awesomeBarObserver.afterInteraction entered",
+      {
+        observedEventsSinceLastFocus: this.observedEventsSinceLastFocus,
+      },
+    ]);
+
+    try {
+      const focusEvent = this.observedEventsSinceLastFocus
+        .slice()
+        .reverse()
+        .find(observedEvent => {
```

Comment:
Oh also not super important but this seems like a common pattern, I wonder if you want a function.

Reply:
True, it did end up being a pattern :) Will wrap in a function in future revisions